From 2b0d4bde3c996dbd5b6fbdce2d986cc3165565f3 Mon Sep 17 00:00:00 2001
From: Lars Francke <git@lars-francke.de>
Date: Wed, 13 Aug 2025 14:16:55 +0200
Subject: NIFI-14858: Make SNI checking configurable

Introduces two new properties:
- nifi.web.https.sni.required (default: false)
- nifi.web.https.sni.host.check (default: true)

These defaults mean that SNI is not required (this is the current behavior already) but if SNI is provided then the host has to match.
---
 .../StandardServerConnectorFactory.java       | 24 +++++++++++++++++++
 .../org/apache/nifi/util/NiFiProperties.java  | 10 ++++++++
 .../FrameworkServerConnectorFactory.java      |  4 ++++
 3 files changed, 38 insertions(+)

diff --git a/nifi-commons/nifi-jetty-configuration/src/main/java/org/apache/nifi/jetty/configuration/connector/StandardServerConnectorFactory.java b/nifi-commons/nifi-jetty-configuration/src/main/java/org/apache/nifi/jetty/configuration/connector/StandardServerConnectorFactory.java
index 892755cdedddda6ea2d989aa0e78f950f4a2e88b..856ef646b01a1f72007cb91c949546d87af1ec88 100644
--- a/nifi-commons/nifi-jetty-configuration/src/main/java/org/apache/nifi/jetty/configuration/connector/StandardServerConnectorFactory.java
+++ b/nifi-commons/nifi-jetty-configuration/src/main/java/org/apache/nifi/jetty/configuration/connector/StandardServerConnectorFactory.java
@@ -70,6 +70,10 @@ public class StandardServerConnectorFactory implements ServerConnectorFactory {
 
     private int requestHeaderSize = 8192;
 
+    private boolean sniRequired = false;
+
+    private boolean sniHostCheck = true;
+
     /**
      * Standard Server Connector Factory Constructor with required properties
      *
@@ -181,6 +185,24 @@ public class StandardServerConnectorFactory implements ServerConnectorFactory {
         this.requestHeaderSize = requestHeaderSize;
     }
 
+    /**
+     * Set to true if a SNI certificate is required, else requests will be rejected with 400 response.
+     *
+     * @param sniRequired SNI Required status
+     */
+    public void setSniRequired(final boolean sniRequired) {
+        this.sniRequired = sniRequired;
+    }
+
+    /**
+     * Set to true if the SNI Host name must match when there is an SNI certificate.
+     *
+     * @param sniHostCheck SNI Host Check status
+     */
+    public void setSniHostCheck(final boolean sniHostCheck) {
+        this.sniHostCheck = sniHostCheck;
+    }
+
     protected Server getServer() {
         return server;
     }
@@ -195,6 +217,8 @@ public class StandardServerConnectorFactory implements ServerConnectorFactory {
             httpConfiguration.setSendServerVersion(SEND_SERVER_VERSION);
 
             final SecureRequestCustomizer secureRequestCustomizer = new SecureRequestCustomizer();
+            secureRequestCustomizer.setSniRequired(sniRequired);
+            secureRequestCustomizer.setSniHostCheck(sniHostCheck);
             httpConfiguration.addCustomizer(secureRequestCustomizer);
         }
 
diff --git a/nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java b/nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java
index 48681c77c705d63e14d5cc165bb26fd96ea94e90..c9763549b14521a3e43d258889b97c3e286812c3 100644
--- a/nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java
+++ b/nifi-commons/nifi-properties/src/main/java/org/apache/nifi/util/NiFiProperties.java
@@ -223,6 +223,8 @@ public class NiFiProperties extends ApplicationProperties {
     public static final String WEB_HTTPS_CIPHERSUITES_INCLUDE = "nifi.web.https.ciphersuites.include";
     public static final String WEB_HTTPS_CIPHERSUITES_EXCLUDE = "nifi.web.https.ciphersuites.exclude";
     public static final String WEB_HTTPS_NETWORK_INTERFACE_PREFIX = "nifi.web.https.network.interface.";
+    public static final String WEB_HTTPS_SNI_REQUIRED = "nifi.web.https.sni.required";
+    public static final String WEB_HTTPS_SNI_HOST_CHECK = "nifi.web.https.sni.host.check";
     public static final String WEB_WORKING_DIR = "nifi.web.jetty.working.directory";
     public static final String WEB_THREADS = "nifi.web.jetty.threads";
     public static final String WEB_MAX_HEADER_SIZE = "nifi.web.max.header.size";
@@ -740,6 +742,14 @@ public class NiFiProperties extends ApplicationProperties {
         return Arrays.stream(protocols.split("\\s+")).collect(Collectors.toSet());
     }
 
+    public boolean isWebHttpsSniRequired() {
+        return Boolean.parseBoolean(getProperty(WEB_HTTPS_SNI_REQUIRED, "false"));
+    }
+
+    public boolean isWebHttpsSniHostCheck() {
+        return Boolean.parseBoolean(getProperty(WEB_HTTPS_SNI_HOST_CHECK, "true"));
+    }
+
     public String getWebMaxHeaderSize() {
         return getProperty(WEB_MAX_HEADER_SIZE, DEFAULT_WEB_MAX_HEADER_SIZE);
     }
diff --git a/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/connector/FrameworkServerConnectorFactory.java b/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/connector/FrameworkServerConnectorFactory.java
index 9d4c5f56a21baa0f49cc789e545dbe1e2852dbbf..820f16ec31eb5059680e2f016fd94d5d8bb182a1 100644
--- a/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/connector/FrameworkServerConnectorFactory.java
+++ b/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/connector/FrameworkServerConnectorFactory.java
@@ -95,6 +95,10 @@ public class FrameworkServerConnectorFactory extends StandardServerConnectorFact
 
             // Set Transport Layer Security Protocols based on platform configuration
             setIncludeSecurityProtocols(TlsPlatform.getPreferredProtocols().toArray(new String[0]));
+
+            // Set SNI configuration from properties
+            setSniRequired(properties.isWebHttpsSniRequired());
+            setSniHostCheck(properties.isWebHttpsSniHostCheck());
         }
     }
 
